optuna/integration/fastai.py

```diff
@@ -0,0 +1,93 @@
+from __future__ import print_function
+
+import optuna
+from optuna import type_checking
+
+if type_checking.TYPE_CHECKING:
+    from typing import Any  # NOQA
+
+try:
+    from fastai.basic_train import Learner  # NOQA
+    from fastai.callbacks import TrackerCallback
+    _available = True
+except ImportError as e:
+    _import_error = e
+    _available = False
+    TrackerCallback = object
+
+
+class FastAIPruningCallback(TrackerCallback):
+    """FastAI callback to prune unpromising trials for fastai.
+
+    .. note::
+        This callback is for fastai<2.0, not the coming version developed in fastai/fastai_dev.
+
+    Example:
+
+        Add a pruning callback which monitors validation loss directly to ``Learner``.
+
+        .. code::
+
+            # If registering this callback in construction
+            from functools import partial
+
+            learn = Learner(
+                data, model,
+                callback_fns=[partial(FastAIPruningCallback, trial=trial, monitor='valid_loss')])
+
+    Example:
+
+        Register a pruning callback to ``learn.fit`` and ``learn.fit_one_cycle``.
+
+        .. code::
+
+            learn.fit(n_epochs, callbacks=[FastAIPruningCallback(learn, trial, 'valid_loss')])
+            learn.fit_one_cycle(
+                n_epochs, cyc_len, max_lr,
+                callbacks=[FastAIPruningCallback(learn, trial, 'valid_loss')])
+
+
+    Args:
+        learn:
+            `fastai.basic_train.Learner <https://docs.fast.ai/basic_train.html#Learner>`_.
+        trial:
+            A :class:`~optuna.trial.Trial` corresponding to the current
+            evaluation of the objective function.
+        monitor:
+            An evaluation metric for pruning, e.g. ``valid_loss`` and ``Accuracy``.
+            Please refer to `fastai.Callback reference
+            <https://docs.fast.ai/callback.html#Callback>`_ for further
+            details.
+    """
+
+    def __init__(self, learn, trial, monitor):
+        # type: (Learner, optuna.trial.Trial, str) -> None
+
+        super(FastAIPruningCallback, self).__init__(learn, monitor)
+
+        _check_fastai_availability()
+
+        self.trial = trial
+
+    def on_epoch_end(self, epoch, **kwargs):
+        # type: (int, Any) -> None
+
+        value = self.get_monitor_value()
+        if value is None:
+            return
+
+        self.trial.report(value, step=epoch)
```

Comment:
Please convert the type of `value`  to `float` because it can be `numpy.ndarray`. If a `numpy.ndarray` value is reported to MySQL, the same issue as #642 and #655 will be occurred.


```suggestion
        self.trial.report(float(value), step=epoch)
```

Reply:
Thanks for pointing it out.

IIUC, it is enough if `value` is int or float.

So, I'd like to use `value.item()` as follows instead of `float(value)` as it's faster and it's clearer for me to understand what the code does.

```python
if not isinstance(value, (int, float)):
    value = value.item()
```

Any thoughts?

Ref:
```shell
In [25]: x, x_tensor                                                                                                                                                                                                                                      
Out[25]: (array([1.]), tensor([1.], dtype=torch.float64))

In [26]: %timeit float(x)                                                                                                                                                                                                                                 
107 ns ± 2.37 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [27]: %timeit x.item()                                                                                                                                                                                                                                 
71.7 ns ± 1.06 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [28]: %timeit float(x_tensor)                                                                                                                                                                                                                          
519 ns ± 10.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [29]: %timeit x_tensor.item()                                                                                                                                                                                                                          
456 ns ± 4.19 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```

Reply:
I think checking `value`'s type every time and calling `item` method conditionally are a bit too much.
Therefore, just `float(value)` looks reasonable to me.`

Reply:
Conclusion: Use `float(value)` 

https://github.com/pfnet/optuna/pull/585/commits/0db16d25441f1657d11a4c38a0cb8996070e9cc8

Reply:
I agree with you. `float(value)` may not be perfect, but it seems to be acceptable.
Thank you for your careful consideration.